presto-main/src/main/java/com/facebook/presto/cost/CostCalculator.java

```diff
@@ -0,0 +1,243 @@
+    {
+        this.numberOfNodes = numberOfNodes;
+    }
+
+    public PlanNodeCostEstimate calculateCost(
+            Session session,
+            Map<Symbol, Type> types,
+            PlanNode planNode,
+            Lookup lookup)
+    {
+        Visitor visitor = new Visitor(
+                session,
+                types,
+                lookup,
+                numberOfNodes);
+
+        PlanNodeCostEstimate childrenCost = planNode.getSources().stream()
+                .map(child -> lookup.getCost(session, types, child))
+                .reduce(PlanNodeCostEstimate.builder().build(), PlanNodeCostEstimate::add);
+        return planNode.accept(visitor, null).add(childrenCost);
+    }
+
+    private class Visitor
+            extends PlanVisitor<Void, PlanNodeCostEstimate>
+    {
+        private final Session session;
+        private final Map<Symbol, Type> types;
+        private final Lookup lookup;
+        private final int numberOfNodes;
+
+        public Visitor(Session session, Map<Symbol, Type> types, Lookup lookup, int numberOfNodes)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.types = requireNonNull(types, "types is null");
+            this.lookup = lookup;
+            this.numberOfNodes = numberOfNodes;
+        }
+
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Void context)
+        {
+            return UNKNOWN_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitFilter(FilterNode node, Void context)
+        {
+            return cpuCost(getStats(node.getSource()).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitProject(ProjectNode node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Void context)
+        {
+            PlanNodeStatsEstimate aggregationStats = getStats(node);
+            return PlanNodeCostEstimate.builder()
+                    .setCpuCost(aggregationStats.getOutputSizeInBytes())
+                    .setMemoryCost(aggregationStats.getOutputSizeInBytes())
+                    .build();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitJoin(JoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node,
+                    node.getLeft(),
+                    node.getRight(),
+                    node.getDistributionType().orElse(JoinNode.DistributionType.PARTITIONED).equals(JoinNode.DistributionType.REPLICATED));
+        }
+
+        private PlanNodeCostEstimate calculateJoinCost(PlanNode join, PlanNode probe, PlanNode build, boolean replicated)
```

Comment:
Could we calculate the network cost here too?  I guess it becomes complicated because once exchanges are added, then you'd be counting the network cost twice.  But otherwise replicated joins will always be more costly than repartitioned.